File: java/cluster_management/src/main/java/org/example/DeleteMultiRegionClusters.java
package org.example;

import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.retries.api.BackoffStrategy;
import software.amazon.awssdk.services.dsql.DsqlClient;
import software.amazon.awssdk.services.dsql.DsqlClientBuilder;

import java.time.Duration;
import java.util.Optional;

public class DeleteMultiRegionClusters {

    public static void main(String[] args) {
        Region region1 = Region.of(System.getenv().getOrDefault("REGION_1", "us-east-1"));
        String clusterId1 = Optional.ofNullable(System.getenv("CLUSTER_ID_1"))
                .orElseThrow(() -> new IllegalStateException("Expected CLUSTER_ID_1 in environment"));
        Region region2 = Region.of(System.getenv().getOrDefault("REGION_2", "us-east-1"));
        String clusterId2 = Optional.ofNullable(System.getenv("CLUSTER_ID_2"))
                .orElseThrow(() -> new IllegalStateException("Expected CLUSTER_ID_2 in environment"));

        DsqlClientBuilder clientBuilder = DsqlClient.builder()
                .credentialsProvider(DefaultCredentialsProvider.create());

        try (
            DsqlClient client1 = clientBuilder.region(region1).build();
            DsqlClient client2 = clientBuilder.region(region2).build()
        ) {
            example(client1, clusterId1, client2, clusterId2);
        }
    }

    public static void example(DsqlClient client1, String clusterId1, DsqlClient client2, String clusterId2) {
        System.out.printf("Deleting cluster %s in %s%n", clusterId1, client1.serviceClientConfiguration().region());
        client1.deleteCluster(r -> r.identifier(clusterId1));

        // cluster1 will stay in PENDING_DELETE until cluster2 is deleted.

        System.out.printf("Deleting cluster %s in %s%n", clusterId2, client2.serviceClientConfiguration().region());
        client2.deleteCluster(r -> r.identifier(clusterId2));

        // Now that both clusters have been marked for deletion they will transition
        // to DELETING state and finalize deletion.
        System.out.printf("Waiting for cluster %s to finish deletion%n", clusterId1);
        client1.waiter().waitUntilClusterNotExists(
                getCluster -> getCluster.identifier(clusterId1),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        );

        System.out.printf("Waiting for cluster %s to finish deletion%n", clusterId2);
        client2.waiter().waitUntilClusterNotExists(
                getCluster -> getCluster.identifier(clusterId2),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        );

        System.out.printf("Deleted %s in %s and %s in %s%n",
                clusterId1, client1.serviceClientConfiguration().region(),
                clusterId2, client2.serviceClientConfiguration().region());
    }
}

File: javascript/cluster_management/src/delete_multi_region_clusters.js
import { DSQLClient, DeleteClusterCommand, waitUntilClusterNotExists } from "@aws-sdk/client-dsql";

export async function deleteMultiRegionClusters(region1, cluster1_id, region2, cluster2_id) {

    const client1 = new DSQLClient({ region: region1 });
    const client2 = new DSQLClient({ region: region2 });

    try {
        const deleteClusterCommand1 = new DeleteClusterCommand({
            identifier: cluster1_id,
        });
        const response1 = await client1.send(deleteClusterCommand1);

        const deleteClusterCommand2 = new DeleteClusterCommand({
            identifier: cluster2_id,
        });
        const response2 = await client2.send(deleteClusterCommand2);

        console.log(`Waiting for cluster1 ${response1.identifier} to finish deletion`);
        await waitUntilClusterNotExists(
            {
                client: client1,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response1.identifier
            }
        );
        console.log(`Cluster1 Id ${response1.identifier} is now deleted`);

        console.log(`Waiting for cluster2 ${response2.identifier} to finish deletion`);
        await waitUntilClusterNotExists(
            {
                client: client2,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response2.identifier
            }
        );
        console.log(`Cluster2 Id ${response2.identifier} is now deleted`);
        return;
    } catch (error) {
        if (error.name === "ResourceNotFoundException") {
            console.log("Some or all Cluster ARNs not found or already deleted");
        } else {
            console.error("Unable to delete multi-region clusters: ", error.message);
        }
        throw error;
    }
}

async function main() {
    const region1 = "us-east-1";
    const cluster1_id = "<CLUSTER_ID_1>";
    const region2 = "us-east-2";
    const cluster2_id = "<CLUSTER_ID_2>";

    const response = await deleteMultiRegionClusters(region1, cluster1_id, region2, cluster2_id);
    console.log(`Deleted ${cluster1_id} in ${region1} and ${cluster2_id} in ${region2}`);
}

if (process.env.NODE_ENV !== 'test') {
    main();
}
